
# Linear and Non-Linear Programming

by Emil Vassev

October 9, October 12, 2022
<br><br>
Copyright (C) 2022 - All rights reserved, do not copy or distribute without permission of the author.
***

## Machine Learning Optimization 
ML optimization is the process of:
* maximizing or minimizing a function by systematically choosing input values from a set, so to compute an optimal value of that function
* adjusting hyperparameters in order to minimize the cost function by using one of the optimization techniques

## Linear Programming vs Nonlinear Programming
**Linear programming** is a mathematical method that allows determining the best possible outcome or solution from a given set of parameters or a set of requirements. These parameters or requirements have a **linear relationship**. On the other hand, **nonlinear programming** is the mathematical method of finding the optimized solution by considering **constraints or objective functions that are nonlinear**.

The main advantage of linear programming is that it helps to perform modeling or simulation to find the best solution according to the available resources, e.g., money, energy, time, space and other related factors or variables. The outcome of linear programming is maximizing or minimizing the result, e.g., maximizing a profit or minimizing costs.


### Generic Components

1. Objective function: maximum or minimum
    * objective function f(x) needs to be either maximized or minimized
    * general minimization problem 
      ```python 
      if max[f(x)] then -f(x)=min[f(x)]
      ```
2. Decision variables - chosen to minimize the function, i.e., `x` that `min[f(x)]`
3. Constraints - constrain decision variables to a certain set of values, e.g.,  `x є [a,b]`

### Specifics
1. Linear programming decision variable is continuous, functional form is linear and constraints are linear.
2. Nonlinear programming decision variable is continuous and either the objective function or constraints are non linear.


## Components of Linear Programming Problem

 $LPP = < D, f, F_{c}, C, C_{n}, R, T_{c} >$
 
* $D$ *decision variables* - to be estimated as an output of the LPP solution
* $f$ *objective function* - evaluates the amount by which decision variables would contribute to $min[f(x)]$ or $max[f(x)]$ 
* $F_{c}$ *objective function coefficients* - amount by which the objective function would change when one decision variable is altered
* $C$ *constraints* -  restrictions or limitations of decision variables
* $C_{n}$ *non-negative constraints* - decision variables must be positive for both maximization or minimization problems (some LPPs can ignore $C_{n}$)
* $R$ *resource availability* - resource boundaries used to define the constraints $C$
* $T_{c}$ *technological coefficients* - coefficients used to define the constraints $C$

### LPP Example 
Maximize `f = 15*x1 + 10*x2`

Subject to:
* `0.25*x1 + 1*x2 ≤ 65`
* `1.25*x1 + 0.5*x2 ≤ 90` 
* `x1 ≥ 0`
* `x2 ≥ 0`

* $D$ *decision variables* - `D={x1, x2}`
* $f$ *objective function* - `max[f] = max[15*x1 + 10 *x2]`
* $F_{c}$ *objective function coefficients* - `Fc = {15:x1, 10:x2}`
* $C$ *constraints* -  `C = {(0.25*x1 + 1*x2 ≤ 65), (1.25*x1 + 0 5*x2 ≤ 90)}`
* $C_{n}$ *non-negative constraints* - `Cn = {x1≥0, x2≥0}`
* $R$ *resource availability* - `R = {65, 90}`
* $T_{c}$ *technological coefficients* - `Tc = {0.25, 1, 1.25, 0.5}`

## Standard Form of LP
A linear program is said to be in standard form iif:
* it is a maximization program
* there are only equalities (no inequalities) in constraint specifications
* all variables are restricted to be nonnegative

### LP Transformation to a Standard Form
1. If objective function is `m𝑖𝑛[𝑧], 𝑧 = 𝑐0 + 𝑐1∗𝑥1 + … + 𝑐𝑛∗𝑥𝑛`
   * maximize it: `max[𝑧], −𝑧 = −𝑐0 − 𝑐1∗𝑥1 − ⋯ − 𝑐𝑛∗𝑥𝑛`
   
2. If there are inequality constraints:
   * if of the form `𝑎1∗𝑥1 + 𝑎2∗𝑥2 + ⋯ + 𝑎𝑛∗𝑥𝑛 ≤ 𝑏`
       * transform it into an equality constraint by adding a slack variable `𝑠`: `𝑎1∗𝑥1+𝑎2∗𝑥2 + ⋯ + 𝑎𝑛∗𝑥𝑛 + 𝑠 = 𝑏, 𝑠≥0`
   * if of the form `𝑎1∗𝑥1 + 𝑎2∗𝑥2 + ⋯ + 𝑎𝑛∗𝑥𝑛 ≥ 𝑏`
       * transform it into an equality constraint by adding a surplus variable `𝑠`: `𝑎1∗𝑥1+𝑎2∗𝑥2 + ⋯ + 𝑎𝑛∗𝑥𝑛 - 𝑠 = 𝑏, 𝑠≥0`   
3. If a decision variable `x` is unrestricted in sign:
   * introduce two new decision variables `𝑥′`and `𝑥′′` where `𝑥 = 𝑥′−𝑥′′`, `𝑥′≥0`, `𝑥′′≥0`
   * replace every occurrence of `𝑥` with `𝑥′−𝑥′′`
   
### Example

Objective function `m𝑖𝑛[𝑧]`, `𝑧 = 2𝑥1 − 𝑥2`

Subject to:
* `𝑥1 + 𝑥2 ≥ 2` 
* `3𝑥1 + 2𝑥2 ≤ 4` 
* `𝑥1 + 𝑥2 = 3`
* `𝑥1 ≶ 0` 
* `𝑥2 ≥ 0`

**Standard form**

`𝑥1 = 𝑥1′ - 𝑥1′′`

Objective function `m𝑎𝑥[𝑧]`, `−𝑧 = −2𝑥1′ + 2𝑥1′′ + 𝑥2`

Subject to:
* `𝑥1′ − 𝑥1′′ + 𝑥2 − 𝑥3 = 2` 
* `3𝑥1′ − 3𝑥1′′ + 2𝑥2 + 𝑥4 = 4`
* `𝑥1′ − 𝑥1′′ + 𝑥2 = 3`
* `𝑥1′ ≥ 0`
* `𝑥1′′ ≥ 0`
* `𝑥2 ≥ 0`
* `𝑥3 ≥ 0`
* `𝑥4 ≥ 0`

## Linear Programming With Python

### Example #1

This example has been extracted from <a link="https://realpython.com/linear-programming-python/" target="new">Hands-On Linear Programming: Optimization With Python</a>

<div>
<img src="images/lpp.png" align="left" />
</div> 

In [1]:
from scipy.optimize import linprog

### Note
The `linprog()` function solves only minimization (not maximization) problems and doesn’t allow inequality constraints with the *greater than or equal to sign* (`≥`). To work around these issues, you need to modify your problem before starting the optimization:

* Instead of maximizing `z = x + 2y`, you can minimize its negative form (`−z = −x − 2y`).
* Instead of having the *greater than or equal to sign*, you can multiply the yellow inequality by `−1` and get the opposite less than or equal to sign (`≤`).

In [2]:
obj = [-1, -2]
#      ─┬  ─┬
#       │   └┤ Coefficient for y
#       └────┤ Coefficient for x

lhs_ineq = [[ 2,  1],  # Red constraint left side
            [-4,  5],  # Blue constraint left side
            [ 1, -2]]  # Yellow constraint left side

rhs_ineq = [20,  # Red constraint right side
            10,  # Blue constraint right side
             2]  # Yellow constraint right side

lhs_eq = [[-1, 5]]  # Green constraint left side
rhs_eq = [15]       # Green constraint right side

In the following lines of code:

* `bnd` definition - defines the bounds for each variable in the same order as the coefficients (non-negative constraints)
* `linprog()` function - optimizes and solves the problem by executing
* `opt` – shows the result

In [3]:
bnd = [(0, float("inf")),  # Bounds of x
       (0, float("inf"))]  # Bounds of y

In [4]:
opt = linprog(c=obj, A_ub=lhs_ineq, b_ub=rhs_ineq,
              A_eq=lhs_eq, b_eq=rhs_eq, bounds=bnd,
              method='highs')

The algorithm used to solve the LPP is **highs** (default). Other supported algorithms are:
* highs-ds
* highs-ipm
* interior-point
* revised simplex
* simplex

https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linprog.html

In [5]:
opt

           con: array([0.])
 crossover_nit: 0
         eqlin:  marginals: array([-0.27272727])
  residual: array([0.])
           fun: -16.818181818181817
       ineqlin:  marginals: array([-0.63636364, -0.        , -0.        ])
  residual: array([ 0.        , 18.18181818,  3.36363636])
         lower:  marginals: array([0., 0.])
  residual: array([7.72727273, 4.54545455])
       message: 'Optimization terminated successfully. (HiGHS Status 7: Optimal)'
           nit: 0
         slack: array([ 0.        , 18.18181818,  3.36363636])
        status: 0
       success: True
         upper:  marginals: array([0., 0.])
  residual: array([inf, inf])
             x: array([7.72727273, 4.54545455])

### Result
* `x = 7.72727273`
* `y = 4.54545455`
* `max[z] = 16.818181818181817`

### Example #2

A company makes two products (X and Y) using two machines (A and B). Each unit of X that is produced requires 50 minutes processing time on machine A and 30 minutes processing time on machine B. Each unit of Y that is produced requires 24 minutes processing time on machine A and 33 minutes processing time on machine B.

At the start of the current week there are 30 units of X and 90 units of Y in stock. Available processing time on machine A is forecast to be 40 hours and on machine B is forecast to be 35 hours.

The demand for X in the current week is forecast to be 75 units and for Y is forecast to be 95 units. Company policy is to maximise the combined sum of the units of X and the units of Y in stock at the end of the week.

Formulate the problem of deciding how much of each product to make in the current week as a linear program.

#### Assumptions 
**Decision variables:**
* `x` - the number of units of X produced in the current week
* `y` - the number of units of Y produced in the current week

**Constraints:**
* `50x + 24y <= 2400` machine A time
* `30x + 33y <= 2100` machine B time
* `x >= 75 - 30`, i.e. x >= 45 so production of X >= demand (75) - initial stock (30), which ensures we meet demand
* `y >= 95 - 90`, i.e. y >= 5 so production of Y >= demand (95) - initial stock (90), which ensures we meet demand

**Objective function:**
* the objective is to maximize the number of units left in stock at the end of the week, i.e., `max[(x+30-75) + (y+90-95)]`
* `max[z], z = x + y - 50`


#### Defining the LPP in Python

**Transformations**:
* Objective function - instead of maximizing `z = x + y - 50`, we will minimize its negative form (`−z = − x − y + 50`).
    * `min[z], -z = -x - y + 50`
    *  our objective function contains a constant (50), but this constant does not change the optimization problem or the optimal solution, i.e., we don't use it in `linprog()`

* Non-negative constraints - the following constraints could be considered as an *upgrade* of the non-negative constraints:
    * `x >= 45`
    * `y >= 5`    

In [6]:
obj = [-1, -1]
#      ─┬  ─┬
#       │   └┤ Coefficient for y
#       └────┤ Coefficient for x

lhs_ineq = [[50,  24],  # 50x + 24y <= 2400
            [30,  33]]  # 30x + 33y <= 2100

rhs_ineq = [2400,  # 50x + 24y <= 2400
            2100]  # 30x + 33y <= 2100


bnd = [(45, float("inf")),  # Bounds of x
       (5, float("inf"))]  # Bounds of y

In [7]:
opt = linprog(c=obj, A_ub=lhs_ineq, b_ub=rhs_ineq,
              bounds=bnd,
              method='highs')
opt

           con: array([], dtype=float64)
 crossover_nit: 0
         eqlin:  marginals: array([], dtype=float64)
  residual: array([], dtype=float64)
           fun: -51.25
       ineqlin:  marginals: array([-0.04166667, -0.        ])
  residual: array([  0.  , 543.75])
         lower:  marginals: array([1.08333333, 0.        ])
  residual: array([0.  , 1.25])
       message: 'Optimization terminated successfully. (HiGHS Status 7: Optimal)'
           nit: 0
         slack: array([  0.  , 543.75])
        status: 0
       success: True
         upper:  marginals: array([0., 0.])
  residual: array([inf, inf])
             x: array([45.  ,  6.25])

### Result
* `x = 45`
* `y = 6.25`
* `max[z] = 51.25 - 50 = 1.25`

## Nonlinear Programming

Nonlinear Programming deals with the problem of optimizing a **nonlinear objective function** subject to nonlinear or linear constraints on the decision variables.

## Components of Nonlinear Programming Problem

 $NLPP = < D, f, F_{c}, C, C_{n}, R, T_{c} >$
 
* $D$ *decision variables* - to be estimated as an output of the NLPP solution
* $f$ *objective function* - evaluates the amount by which decision variables would contribute to $min[f(x)]$  or $max[f(x)]$   
* $F_{c}$ *objective function coefficients* - amount by which the objective function would change when one decision variable is altered
* $C$ *constraints* -  restrictions or limitations of decision variables
* $C_{n}$ *non-negative constraints* - decision variables must be positive for both maximization or minimization problems (not a mandatory rule for NLPPs)
* $R$ *resource availability* - resource boundaries used to define the constraints $C$
* $T_{c}$ *technological coefficients* - coefficients used to define the constraints $C$

### NLPP Example 
Maximize `m𝑎𝑥[𝑧], 𝑧 = 2∗𝑥1 + 𝑥2 − 5∗𝑙𝑜𝑔(𝑥1)∗sin(𝑥2)`

Subject to:
* `𝑥1 ∗ 𝑥2 ≤ 10`
* `|𝑥1 − 𝑥2| ≤ 2`
* `0.1 ≤ 𝑥1 ≤ 5`
* `0.1 ≤ 𝑥2 ≤ 3`

* $D$ *decision variables* - `D={x1, x2}`
* $f$ *objective function* - `max[z], max[2∗𝑥1 + 𝑥2 − 5∗𝑙𝑜𝑔(𝑥1)∗sin(𝑥2)]`
* $F_{c}$ *objective function coefficients* - `Fc = {2:x1, 1:x2, 5:log(x1)∗sin(x2)}`
* $C$ *constraints* -  `C = {(x1∗x2≤10),(|𝑥1−𝑥2|≤2), (0.1≤𝑥1≤5), (0.1≤𝑥2≤3)}`
* $C_{n}$ *non-negative constraints* - `Cn = {x1≥0, x2≥0}`
* $R$ *resource availability* - `R = {10, 2, 0.1, 5, 0.1, 3}`
* $T_{c}$ *technological coefficients* - `Tc = {1, 1, 1, -1, 1, 1}`

## Nonlinear Programming With Python

Objective function:
* $min[z]$, $z=(x_{1}-1)^{2} + (x_{2}-2.5)^{2}$

Subject to:
* $x_{1} - 2*x_{2} >= -2$
* $-x_{1} - 2*x_{2} >= -6$
* $-x_{1} + 2*x_{2} >= -2$
* $x_{1} >= 0$
* $x_{2} >= 0$

We use `scipy.optimize.minimize` function - works on the minimization of a scalar function of one or more variables. There is no `maximize` function.

In [8]:
import numpy as np
from scipy.optimize import minimize

In the following lines of code:

<li><b>obj_fun</b>: defines objective function; decision variables x are an array x[]</li>
<li><b>A</b>: array of constraint coefficients</li>
<li><b>b</b>: array of resource availabilities</li>
<li><b>bonds</b>: defines the bounderies of the decision variables, e.g., non-negative constraints for x1 an x2</li>
<li><b>xinit</b>: starting point of search</li>

In [9]:
obj_fun = lambda x: (x[0] - 1)**2 + (x[1] - 2.5)**2
A = np.array([[1, -2], [-1, -2], [-1, 2]])
b = np.array([-2, -6, -2])
bnds = [(0, None) for i in range(A.shape[1])]  # x_1 >= 0, x_2 >= 0
xinit = [0, 0] 

Constraints: 
* in this case, they are linear and we express them as an affine-linear function `A*x-b`
* defined as a dictionary where `fun` is a callable function such that `fun >= 0`

In [10]:
cons = [{'type': 'ineq', 'fun': lambda x:  x[0] - 2 * x[1] + 2},
        {'type': 'ineq', 'fun': lambda x: -x[0] - 2 * x[1] + 6},
        {'type': 'ineq', 'fun': lambda x: -x[0] + 2 * x[1] + 2}]  #−𝑥1+2∗𝑥2>=−2 --> −𝑥1 + 2 ∗𝑥2 + 2 >= 0

When we define the constraints, the type could be either `ineq` or `eq`:
```python
cons = [{'type': 'ineq', 'fun': lambda x:  x[0] - 2 * x[1] + 2},
        {'type': 'eq', 'fun': lambda x: -x[0] - 2 * x[1] + 6}]         
```
Equality constraint means that the constraint function result is to be zero whereas inequality means that it is to be non-negative.

In [11]:
res = minimize(obj_fun, x0=xinit, bounds=bnds, constraints=cons)
res

     fun: 0.799999999999998
     jac: array([ 0.79999999, -1.59999999])
 message: 'Optimization terminated successfully'
    nfev: 12
     nit: 4
    njev: 4
  status: 0
 success: True
       x: array([1.39999999, 1.69999999])

### Result
* `x = 1.39999999`
* `y = 1.69999999`
* `min[z] = 0.799999999999998`

### What is the method that has been used?

Available methods (solvers):
https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html#scipy.optimize.minimize

* Nelder-Mead
* Powell
* G
* BFGS
* Newton-CG
* L-BFGS-B
* TNC
* COBYLA
* SLSQP
* trust-constr
* dogleg
* trust-ncg
* trust-exact
* trust-krylov

List of parameters:
```python
def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
             hessp=None, bounds=None, constraints=(), tol=None,
             callback=None, options=None)
```

To extract the method that has been used: 

In [12]:
method = None
constraints=cons
bounds=bnds

if method is None:
    # Select automatically
    if constraints:
        method = 'SLSQP'
    elif bounds is not None:
        method = 'L-BFGS-B'
    else:
        method = 'BFGS'

print(method)

SLSQP


Let's try another method.

In [13]:
res = minimize(obj_fun, method='COBYLA', x0=xinit, constraints=cons)
res

     fun: 0.8000000054738174
   maxcv: 0.0
 message: 'Optimization terminated successfully.'
    nfev: 35
  status: 1
 success: True
       x: array([1.40006617, 1.70003309])

Note that we removed the bounds, because COBYLA does not support bounds.